<a href="https://colab.research.google.com/github/sdev138/LungCancerPredictionModel/blob/main/00_tensorflow_fundamentals.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#In this notebook, we're going to cover the fundamental concepts of Tensors using tensoFlow 
#More specifically, we're going to cover: 
  * Introduction to Tensors 
  * Getting information from tensors
  * Manipulating Tensors
  * Tensors & NumPy
  * Using @tf.function (a way to speed up your regular Python functions)
  * Using GPUs with Tensorflow (for TPU's)
  * Excersise to try 


In [1]:
# import TensorFlow 
import tensorflow as tf
print(tf.__version__) 

2.7.0


In [ ]:
# Create tensors with tf.constant() 
scalar = tf.constant(7) 
scalar

<tf.Tensor: shape=(), dtype=int32, numpy=7>

In [ ]:
# Check the number of dimensions of a tensor (ndim stands for number of dimensions)
scalar.ndim

0

In [ ]:
vector = tf.constant([0,10])
vector

<tf.Tensor: shape=(2,), dtype=int32, numpy=array([ 0, 10], dtype=int32)>

In [ ]:
# Check the dimension of the vector
vector.ndim

1

In [ ]:
# Create a matrix (has more than 1 dimension)
matrix = tf.constant([[10,7],[7,10]])
matrix

<tf.Tensor: shape=(2, 2), dtype=int32, numpy=
array([[10,  7],
       [ 7, 10]], dtype=int32)>

In [ ]:
matrix.ndim

2

In [ ]:
matrix2 = tf.constant([[10,7],[7,10],[8,9]])
matrix2

<tf.Tensor: shape=(3, 2), dtype=int32, numpy=
array([[10,  7],
       [ 7, 10],
       [ 8,  9]], dtype=int32)>

In [ ]:
matrix2.ndim

2

In [ ]:
# Create another matrix 
another_matrix = tf.constant([[10., 7.],[3.,2.],[8.,9.]], dtype = tf.float16)
another_matrix


<tf.Tensor: shape=(3, 2), dtype=float16, numpy=
array([[10.,  7.],
       [ 3.,  2.],
       [ 8.,  9.]], dtype=float16)>

In [ ]:
# What's the number of dimensions of another_matrix?
another_matrix.ndim

2

In [ ]:
# The total number of dimensions is equivalent to the number of elements in a shape

In [ ]:
tensor = tf.constant([[[1,2,3],
                       [4,5,6]],
                      [[7,8,9],
                       [10,11,12]],
                      [[13,14,15],
                       [16,17,18]]])
tensor

<tf.Tensor: shape=(3, 2, 3), dtype=int32, numpy=
array([[[ 1,  2,  3],
        [ 4,  5,  6]],

       [[ 7,  8,  9],
        [10, 11, 12]],

       [[13, 14, 15],
        [16, 17, 18]]], dtype=int32)>

In [ ]:
tensor.ndim

3

What we've created so far:

  * Scalar: a single number 
  * Vector: a number with direction (e.g. wind speed and direction)
  * Matrix: a single 2-dimensional array of numbers 
  * Tensor: an n-dimensional array of numbers (where n can be any number, a 0-dimensional tensor is a scalar, a 1-dimensional tensor is a vector) 
  

Making Tensors through variables 

In [ ]:
# Create the same tensor with tf.Variable() as above 
changeable_tensor = tf.Variable([10,7])
unchangeable_tensor = tf.constant([10,7])
changeable_tensor, unchangeable_tensor

(<tf.Variable 'Variable:0' shape=(2,) dtype=int32, numpy=array([10,  7], dtype=int32)>,
 <tf.Tensor: shape=(2,), dtype=int32, numpy=array([10,  7], dtype=int32)>)

In [ ]:
# Let's try change one of the elements in our changeable tensor
changeable_tensor[0] = 7
changeable_tensor

TypeError: ignored

In [ ]:
# How about we try .assign()
changeable_tensor[0].assign(7)
changeable_tensor

<tf.Variable 'Variable:0' shape=(2,) dtype=int32, numpy=array([7, 7], dtype=int32)>

In [ ]:
# What about trying to change our unchangeable tensor
unchangeable_tensor[0].assign(7)
unchangeable_tensor

AttributeError: ignored

In [ ]:
changeable_tensor[1].assign(8)
changeable_tensor

<tf.Variable 'Variable:0' shape=(2,) dtype=int32, numpy=array([7, 8], dtype=int32)>

Note: Rarely in practice will you need to decide whether to use tf.constant or tf.Variable to create tensors, as Tensorflow does this for you. However, if in doubt, use tf.constant and change it later if needed

Creating Random Tensors

Random tensors are tensors of some arbitrary size which contain random numbers



In [ ]:
# Create two random (but the same) tensors
random_1 = tf.random.Generator.from_seed(42) # set seed for reproducibility 
random_1 = random_1.normal(shape=(3,2))
random_2 = tf.random.Generator.from_seed(41)
random_2 = random_2.normal(shape=(3,2))

# Are they equal?
random_1, random_2, random_1 == random_2

(<tf.Tensor: shape=(3, 2), dtype=float32, numpy=
 array([[-0.7565803 , -0.06854702],
        [ 0.07595026, -1.2573844 ],
        [-0.23193763, -1.8107855 ]], dtype=float32)>,
 <tf.Tensor: shape=(3, 2), dtype=float32, numpy=
 array([[ 0.31662667, -1.4391748 ],
        [ 0.58923835, -1.4268045 ],
        [-0.7565803 , -0.06854702]], dtype=float32)>,
 <tf.Tensor: shape=(3, 2), dtype=bool, numpy=
 array([[False, False],
        [False, False],
        [False, False]])>)

In [ ]:
#creating more random tensors
random_3 = tf.random.Generator.from_seed(38) 
random_3 = random_3.normal(shape=(4,1))
random_4 = tf.random.Generator.from_seed(38)
random_4 = random_4.normal(shape = (4,1))

random_3, random_4, random_3 == random_4

(<tf.Tensor: shape=(4, 1), dtype=float32, numpy=
 array([[ 0.20386009],
        [ 0.562024  ],
        [-2.3001142 ],
        [-1.349454  ]], dtype=float32)>,
 <tf.Tensor: shape=(4, 1), dtype=float32, numpy=
 array([[ 0.20386009],
        [ 0.562024  ],
        [-2.3001142 ],
        [-1.349454  ]], dtype=float32)>,
 <tf.Tensor: shape=(4, 1), dtype=bool, numpy=
 array([[ True],
        [ True],
        [ True],
        [ True]])>)

# ***Shuffling the Order of Tensors***

In [3]:
#Shuffle a tensor (valuable for when you want to shuffle your data so the inherent order doesn't affect learning)
not_shuffled = tf.constant([[10,7],
                           [3,4],
                           [2,5]]) 
not_shuffled.ndim

2

In [4]:
not_shuffled

<tf.Tensor: shape=(3, 2), dtype=int32, numpy=
array([[10,  7],
       [ 3,  4],
       [ 2,  5]], dtype=int32)>

In [5]:

tf.random.shuffle(
    not_shuffled
)

<tf.Tensor: shape=(3, 2), dtype=int32, numpy=
array([[ 3,  4],
       [ 2,  5],
       [10,  7]], dtype=int32)>

In [6]:
# Shuffle our non-shuffled tensor
tf.random.set_seed(40)
tf.random.shuffle(not_shuffled, seed=40)




<tf.Tensor: shape=(3, 2), dtype=int32, numpy=
array([[ 2,  5],
       [10,  7],
       [ 3,  4]], dtype=int32)>

In [7]:
tf.random.set_seed(39)
tf.random.shuffle(not_shuffled, seed = 39)

<tf.Tensor: shape=(3, 2), dtype=int32, numpy=
array([[10,  7],
       [ 3,  4],
       [ 2,  5]], dtype=int32)>

# ***Read the Tensorflow Documentation https://www.tensorflow.org/api_docs/python/tf/random/set_seed***

to use the shuffled tensors to be in the same order, we've got to use the global level random seed as well as the operation level random seed

In [9]:
tf.random.set_seed(42) # global level random seed 
tf.random.shuffle(not_shuffled, seed = 42) # operation level random seed

<tf.Tensor: shape=(3, 2), dtype=int32, numpy=
array([[10,  7],
       [ 3,  4],
       [ 2,  5]], dtype=int32)>

In [ ]:
#Other ways to make tensors


In [10]:
#Create a tensor of all ones 
tf.ones([10,7])

<tf.Tensor: shape=(10, 7), dtype=float32, numpy=
array([[1., 1., 1., 1., 1., 1., 1.],
       [1., 1., 1., 1., 1., 1., 1.],
       [1., 1., 1., 1., 1., 1., 1.],
       [1., 1., 1., 1., 1., 1., 1.],
       [1., 1., 1., 1., 1., 1., 1.],
       [1., 1., 1., 1., 1., 1., 1.],
       [1., 1., 1., 1., 1., 1., 1.],
       [1., 1., 1., 1., 1., 1., 1.],
       [1., 1., 1., 1., 1., 1., 1.],
       [1., 1., 1., 1., 1., 1., 1.]], dtype=float32)>

In [11]:
#Create a tensor of all zeros
tf.zeros(shape = (3,4))

<tf.Tensor: shape=(3, 4), dtype=float32, numpy=
array([[0., 0., 0., 0.],
       [0., 0., 0., 0.],
       [0., 0., 0., 0.]], dtype=float32)>

In [12]:
tf.zeros([4,4])

<tf.Tensor: shape=(4, 4), dtype=float32, numpy=
array([[0., 0., 0., 0.],
       [0., 0., 0., 0.],
       [0., 0., 0., 0.],
       [0., 0., 0., 0.]], dtype=float32)>

# ***Turn numpy arrays into tensors***

The main difference between NumPy arrays and TensorFlow tensors is that tensors can be run on a GPU (much faster for numerical computing). 


In [14]:
# You can also turn NumPy arrays into tensors
import numpy as np 

numpy_A = np.arange(1,25, dtype = np.int32) # Create a numpy array between 1 and 25 

# X = tf.constant(some_matrix) # capital for matrix or tensor 
# y = tf.constant(vector) # non-capital for vector 

numpy_A

array([ 1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14, 15, 16, 17,
       18, 19, 20, 21, 22, 23, 24], dtype=int32)

We got a tensor by passing the numpy array into tf.constant

In [17]:
A = tf.constant(numpy_A, shape = (2,3,4))
A

<tf.Tensor: shape=(2, 3, 4), dtype=int32, numpy=
array([[[ 1,  2,  3,  4],
        [ 5,  6,  7,  8],
        [ 9, 10, 11, 12]],

       [[13, 14, 15, 16],
        [17, 18, 19, 20],
        [21, 22, 23, 24]]], dtype=int32)>

In [18]:
B = tf.constant(numpy_A)
B

<tf.Tensor: shape=(24,), dtype=int32, numpy=
array([ 1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14, 15, 16, 17,
       18, 19, 20, 21, 22, 23, 24], dtype=int32)>

In [20]:
C = tf.constant(numpy_A, shape = (3,8))
C

<tf.Tensor: shape=(3, 8), dtype=int32, numpy=
array([[ 1,  2,  3,  4,  5,  6,  7,  8],
       [ 9, 10, 11, 12, 13, 14, 15, 16],
       [17, 18, 19, 20, 21, 22, 23, 24]], dtype=int32)>

In [21]:
A.ndim

3

# **Getting Information from Tensors**

When dealing with tensors you must be aware of the following attributes: 


*   Shape: The length (number of elements) of each of the dimensions of a tensor (tensor.shape)
*   Rank: The number of tensor dimensions. A scalar has rank 0, a vector has rank1, a matrix is rank 2, a tensor has rank n (tensor.ndim) 
*   Axis or Dimension: A particular dimension of a tensor (tensor[0], tensor[:,1]..) 
*   Size: The total number of items in the tensor (tf.size(tensor)) 





In [22]:
# Create a rank 4 tensor (4 dimensions)

rank_4_tensor = tf.zeros(shape = [2,3,4,5]) 
rank_4_tensor

<tf.Tensor: shape=(2, 3, 4, 5), dtype=float32, numpy=
array([[[[0., 0., 0., 0., 0.],
         [0., 0., 0., 0., 0.],
         [0., 0., 0., 0., 0.],
         [0., 0., 0., 0., 0.]],

        [[0., 0., 0., 0., 0.],
         [0., 0., 0., 0., 0.],
         [0., 0., 0., 0., 0.],
         [0., 0., 0., 0., 0.]],

        [[0., 0., 0., 0., 0.],
         [0., 0., 0., 0., 0.],
         [0., 0., 0., 0., 0.],
         [0., 0., 0., 0., 0.]]],


       [[[0., 0., 0., 0., 0.],
         [0., 0., 0., 0., 0.],
         [0., 0., 0., 0., 0.],
         [0., 0., 0., 0., 0.]],

        [[0., 0., 0., 0., 0.],
         [0., 0., 0., 0., 0.],
         [0., 0., 0., 0., 0.],
         [0., 0., 0., 0., 0.]],

        [[0., 0., 0., 0., 0.],
         [0., 0., 0., 0., 0.],
         [0., 0., 0., 0., 0.],
         [0., 0., 0., 0., 0.]]]], dtype=float32)>

In [23]:
#Asking for its zeroth element which is the first element of its two 
rank_4_tensor[0]

<tf.Tensor: shape=(3, 4, 5), dtype=float32, numpy=
array([[[0., 0., 0., 0., 0.],
        [0., 0., 0., 0., 0.],
        [0., 0., 0., 0., 0.],
        [0., 0., 0., 0., 0.]],

       [[0., 0., 0., 0., 0.],
        [0., 0., 0., 0., 0.],
        [0., 0., 0., 0., 0.],
        [0., 0., 0., 0., 0.]],

       [[0., 0., 0., 0., 0.],
        [0., 0., 0., 0., 0.],
        [0., 0., 0., 0., 0.],
        [0., 0., 0., 0., 0.]]], dtype=float32)>

In [25]:
rank_4_tensor.shape, rank_4_tensor.ndim, tf.size(rank_4_tensor)

(TensorShape([2, 3, 4, 5]), 4, <tf.Tensor: shape=(), dtype=int32, numpy=120>)

[2,3,4,5] is the size of the tensor. Number of dimensions is 4. Size is 120 elements. Because 2*3*4*5 = 120

In [29]:
# Get various attributes of our tensor 

print("Datatype of every element: ", rank_4_tensor.dtype)
print("Number of dimensions (rank): ", rank_4_tensor.ndim) 
print("Shape of Tensor: ", rank_4_tensor.shape)
print("Elements along the 0 axis: ", rank_4_tensor.shape[0])
print("Elements along the last axis: ", rank_4_tensor.shape[-1])
print("The total number of elements in our tensor: ", tf.size(rank_4_tensor))
print("The total number of elements in our tensor: ", tf.size(rank_4_tensor).numpy)

Datatype of every element:  <dtype: 'float32'>
Number of dimensions (rank):  4
Shape of Tensor:  (2, 3, 4, 5)
Elements along the 0 axis:  2
Elements along the last axis:  5
The total number of elements in our tensor:  tf.Tensor(120, shape=(), dtype=int32)
The total number of elements in our tensor:  <bound method _EagerTensorBase.numpy of <tf.Tensor: shape=(), dtype=int32, numpy=120>>


# **Indexing Tensors**

Tensors can be indexed just like Python lists

In [31]:
some_list = [1,2,3,4]
some_list

[1, 2, 3, 4]

In [30]:
# Get the first 2 elements of each dimension 
rank_4_tensor[:2, :2, :2, :2]

<tf.Tensor: shape=(2, 2, 2, 2), dtype=float32, numpy=
array([[[[0., 0.],
         [0., 0.]],

        [[0., 0.],
         [0., 0.]]],


       [[[0., 0.],
         [0., 0.]],

        [[0., 0.],
         [0., 0.]]]], dtype=float32)>

In [35]:
rank_4_tensor.shape

TensorShape([2, 3, 4, 5])

In [37]:
# Get the first element from each dimension from each index except for the final one 

rank_4_tensor[:1, :1, :1, :1] 

<tf.Tensor: shape=(1, 1, 1, 1), dtype=float32, numpy=array([[[[0.]]]], dtype=float32)>

In [39]:
# Create a rank 2 tensor (2 dimensions) 
rank_2_tensor = tf.constant([[10, 7], 
                             [3,4]])
rank_2_tensor.shape, rank_2_tensor.ndim

(TensorShape([2, 2]), 2)

In [40]:
rank_2_tensor

<tf.Tensor: shape=(2, 2), dtype=int32, numpy=
array([[10,  7],
       [ 3,  4]], dtype=int32)>

In [41]:
some_list, some_list[-1]

([1, 2, 3, 4], 4)

In [42]:
# Get the last item of each of row of our rank 2 tensor 
rank_2_tensor[:, -1]

<tf.Tensor: shape=(2,), dtype=int32, numpy=array([7, 4], dtype=int32)>

In [43]:
# Add in extra dimension to our rank 2 tensor 
rank_3_tensor = rank_2_tensor[..., tf.newaxis]
rank_3_tensor

<tf.Tensor: shape=(2, 2, 1), dtype=int32, numpy=
array([[[10],
        [ 7]],

       [[ 3],
        [ 4]]], dtype=int32)>

In [48]:
tensor1 = tf.constant(8,3)
tensor1

<tf.Tensor: shape=(), dtype=int32, numpy=8>

In [51]:
tensor2 = tf.random.Generator.from_seed(69)
tensor2 = tensor2.normal(shape = (3,3))
tensor2

<tf.Tensor: shape=(3, 3), dtype=float32, numpy=
array([[ 0.29164317,  1.4531525 , -0.8223833 ],
       [-1.3446563 , -0.7183838 , -0.20373915],
       [ 0.6291725 , -0.87623316, -0.5923522 ]], dtype=float32)>

In [53]:
tf.expand_dims(rank_2_tensor, axis = -1) # "-1" means expand the final axis 

<tf.Tensor: shape=(2, 2, 1), dtype=int32, numpy=
array([[[10],
        [ 7]],

       [[ 3],
        [ 4]]], dtype=int32)>

In [55]:
tf.expand_dims(rank_2_tensor, axis = 0)

<tf.Tensor: shape=(1, 2, 2), dtype=int32, numpy=
array([[[10,  7],
        [ 3,  4]]], dtype=int32)>

In [56]:
rank_2_tensor

<tf.Tensor: shape=(2, 2), dtype=int32, numpy=
array([[10,  7],
       [ 3,  4]], dtype=int32)>

# **Manipulating Tensors (Tensor operations)**





In [5]:
# + - * /
# you can add values to a tensor using the addition operator 
tensor = tf.constant([[10,7],[3,4]])
tensor + 10

<tf.Tensor: shape=(2, 2), dtype=int32, numpy=
array([[20, 17],
       [13, 14]], dtype=int32)>

In [6]:
# Original tensor is unchanged 
tensor = tensor + 10
tensor 

<tf.Tensor: shape=(2, 2), dtype=int32, numpy=
array([[20, 17],
       [13, 14]], dtype=int32)>

In [7]:
# Multiplication also works 
tensor * 10

<tf.Tensor: shape=(2, 2), dtype=int32, numpy=
array([[200, 170],
       [130, 140]], dtype=int32)>

In [9]:
# Subtraction 
tensor - 20

<tf.Tensor: shape=(2, 2), dtype=int32, numpy=
array([[ 0, -3],
       [-7, -6]], dtype=int32)>

In [13]:
# we can use the tensorflow built in function too 
multiply_tensor = tf.multiply(tensor, 10) # typically the gpu speeds up the calculation if you use the tensorflow built in library 
multiply_tensor

<tf.Tensor: shape=(2, 2), dtype=int32, numpy=
array([[200, 170],
       [130, 140]], dtype=int32)>

In [11]:
tensor 

<tf.Tensor: shape=(2, 2), dtype=int32, numpy=
array([[20, 17],
       [13, 14]], dtype=int32)>

In [14]:
multiply_tensor

<tf.Tensor: shape=(2, 2), dtype=int32, numpy=
array([[200, 170],
       [130, 140]], dtype=int32)>

In [15]:
# In tensorflow machine learning, tensor multiplication is the most used operation for machine learning 
# what is matrix multiplication 
tensor_multiply_2 = tf.constant([[4,4], [8,8]])
multiplication_sum = tensor_multiply_2 * tensor 
multiplication_sum

<tf.Tensor: shape=(2, 2), dtype=int32, numpy=
array([[ 80,  68],
       [104, 112]], dtype=int32)>

In [17]:
# Matrix multiplication in tensorflow 
print(tensor)
tf.matmul(tensor, tensor)

tf.Tensor(
[[20 17]
 [13 14]], shape=(2, 2), dtype=int32)


<tf.Tensor: shape=(2, 2), dtype=int32, numpy=
array([[621, 578],
       [442, 417]], dtype=int32)>

In [20]:
# Matrix multiplication with Python operator "@" 
tensors = tf.constant([[4,3],[5,1]])
tensors

<tf.Tensor: shape=(2, 2), dtype=int32, numpy=
array([[4, 3],
       [5, 1]], dtype=int32)>

In [8]:
# Create a tensor (3, 2) tensor 
X = tf.constant([[1,2], 
                 [3,4], 
                 [5,6]])

#create another (2,2) tensor 
y = tf.constant([[3,2],
                 [2,4]])
tf.matmul(X,y)

<tf.Tensor: shape=(3, 2), dtype=int32, numpy=
array([[ 7, 10],
       [17, 22],
       [27, 34]], dtype=int32)>

In [21]:
X = tf.constant([[1,1],[2,2]])
y = tf.constant([[3,3,3],[4,4,4]])

tf.matmul(X,y)

<tf.Tensor: shape=(2, 3), dtype=int32, numpy=
array([[ 7,  7,  7],
       [14, 14, 14]], dtype=int32)>

In [3]:
Y = tf.random.Generator.from_seed(1022)
Y = Y.normal(shape = (3,2))
Y

<tf.Tensor: shape=(3, 2), dtype=float32, numpy=
array([[-0.72061694,  0.07151537],
       [ 0.5087514 , -0.00330423],
       [ 0.7712984 , -0.5793761 ]], dtype=float32)>

In [38]:
tf.reshape(y, shape = (2,3))

<tf.Tensor: shape=(2, 3), dtype=int32, numpy=
array([[3, 3, 3],
       [4, 4, 4]], dtype=int32)>

In [39]:
X.shape, tf.reshape(y, shape = (2,3)).shape

(TensorShape([2, 3]), TensorShape([2, 3]))

In [40]:
tf.cast(Y, tf.int32)

<tf.Tensor: shape=(3, 2), dtype=int32, numpy=
array([[0, 0],
       [0, 0],
       [0, 0]], dtype=int32)>

In [42]:
# try to matrix multiply X by reshaped y
X @ tf.reshape(y, shape = (2,3))

InvalidArgumentError: ignored

In [37]:
tf.matmul(X, tf.reshape(y, shape = (2,3)))


InvalidArgumentError: ignored

In [33]:
tf.reshape(X, shape=(2,3)).shape, y.shape

InvalidArgumentError: ignored

In [34]:
#try to change the shape of X instead of Y
tf.matmul(tf.reshape(X, shape = (2,3)), y)

InvalidArgumentError: ignored

In [43]:
X, tf.transpose(X), tf.reshape(X, shape = (2,3))

(<tf.Tensor: shape=(2, 3), dtype=int32, numpy=
 array([[ 7,  7,  7],
        [14, 14, 14]], dtype=int32)>,
 <tf.Tensor: shape=(3, 2), dtype=int32, numpy=
 array([[ 7, 14],
        [ 7, 14],
        [ 7, 14]], dtype=int32)>,
 <tf.Tensor: shape=(2, 3), dtype=int32, numpy=
 array([[ 7,  7,  7],
        [14, 14, 14]], dtype=int32)>)

In [45]:
# Try matrix multiplication with transpose rather than reshape 
tf.matmul(tf.transpose(X), y)

<tf.Tensor: shape=(3, 3), dtype=int32, numpy=
array([[77, 77, 77],
       [77, 77, 77],
       [77, 77, 77]], dtype=int32)>

In [46]:
X,Y

(<tf.Tensor: shape=(2, 3), dtype=int32, numpy=
 array([[ 7,  7,  7],
        [14, 14, 14]], dtype=int32)>,
 <tf.Tensor: shape=(3, 2), dtype=float32, numpy=
 array([[-0.72061694,  0.07151537],
        [ 0.5087514 , -0.00330423],
        [ 0.7712984 , -0.5793761 ]], dtype=float32)>)

In [56]:
Y = tf.cast(Y, tf.int32)
Y

<tf.Tensor: shape=(3, 2), dtype=int32, numpy=
array([[0, 0],
       [0, 0],
       [0, 0]], dtype=int32)>

In [58]:
X,Y

(<tf.Tensor: shape=(2, 3), dtype=int32, numpy=
 array([[ 7,  7,  7],
        [14, 14, 14]], dtype=int32)>,
 <tf.Tensor: shape=(3, 2), dtype=int32, numpy=
 array([[0, 0],
        [0, 0],
        [0, 0]], dtype=int32)>)

In [61]:
X.shape,Y.shape

(TensorShape([2, 3]), TensorShape([3, 2]))

In [63]:
# Matrix multiplication is also reffered to as the dot product. You can perform matrix multiplication using:
# tf.matmul()
# tf.tensordot() 
# perform the dot product on X and Y (requires X or Y to be transposed) 
tf.tensordot(X,Y,axes = 1)

<tf.Tensor: shape=(2, 2), dtype=int32, numpy=
array([[0, 0],
       [0, 0]], dtype=int32)>

In [66]:
X.shape, Y.shape

(TensorShape([2, 3]), TensorShape([3, 2]))

In [67]:
tf.tensordot(tf.transpose(X), tf.transpose(Y), axes = 1)

<tf.Tensor: shape=(3, 3), dtype=int32, numpy=
array([[0, 0, 0],
       [0, 0, 0],
       [0, 0, 0]], dtype=int32)>

In [69]:
Y = Y + 10
Y

<tf.Tensor: shape=(3, 2), dtype=int32, numpy=
array([[10, 10],
       [10, 10],
       [10, 10]], dtype=int32)>

In [70]:
# Check the values of Y, reshape Y and transposed Y
print("Normal Y: ")
print(Y, "\n")

print("Y reshaped to (2,3): ")
print(tf.reshape(Y, (2,3)), "\n")

print("Y transposed: ")
print(tf.transpose(Y))

Normal Y: 
tf.Tensor(
[[10 10]
 [10 10]
 [10 10]], shape=(3, 2), dtype=int32) 

Y reshaped to (2,3): 
tf.Tensor(
[[10 10 10]
 [10 10 10]], shape=(2, 3), dtype=int32) 

Y transposed: 
tf.Tensor(
[[10 10 10]
 [10 10 10]], shape=(2, 3), dtype=int32)


# **Changing the datatype of a tensor**

In [93]:
import numpy as np

In [71]:
# Create a new tensor with default datatype (float32)
B = tf.constant([1.7, 7.4])
B.dtype

tf.float32

In [74]:
C = tf.constant([7,10])
C.dtype

tf.int32

In [ ]:
# change from float 32 to float 16 (reduced precision) 

In [76]:
D = tf.cast(B, dtype = tf.float16) 
D, D.dtype

(<tf.Tensor: shape=(2,), dtype=float16, numpy=array([1.7, 7.4], dtype=float16)>,
 tf.float16)

In [77]:
# change from int 32 to float32
E = tf.cast(C, dtype = tf.float32) 
E

<tf.Tensor: shape=(2,), dtype=float32, numpy=array([ 7., 10.], dtype=float32)>

In [78]:
E_float16 = tf.cast(E, dtype = tf.float16)
E_float16

<tf.Tensor: shape=(2,), dtype=float16, numpy=array([ 7., 10.], dtype=float16)>

# **Aggregating Tensors**

Aggregating Tensors = condensing them from multiple values down to a smaller amount of values 

In [79]:
# Getting the absolute values 
D = tf.constant([-7,-10])
D

<tf.Tensor: shape=(2,), dtype=int32, numpy=array([ -7, -10], dtype=int32)>

In [80]:
tf.abs(D)

<tf.Tensor: shape=(2,), dtype=int32, numpy=array([ 7, 10], dtype=int32)>

let's go through the following forms of aggregation:


*   Get the minimum
*   get the max 
*   get the mean of a tensor 
*   get the sum of a tensor 





In [84]:
mean_of_tensor = tf.math.reduce_mean(D, axis = None, keepdims = False, name = None)
mean_of_tensor

<tf.Tensor: shape=(), dtype=int32, numpy=-8>

In [85]:
mean_of_tensor.shape

TensorShape([])

In [88]:
min_of_tensor = tf.math.reduce_min(D, axis = None, keepdims = False, name = None)
min_of_tensor 

<tf.Tensor: shape=(), dtype=int32, numpy=-10>

In [89]:
max_of_tensor = tf.math.reduce_max(D, axis = None, keepdims = False, name = None)
max_of_tensor

<tf.Tensor: shape=(), dtype=int32, numpy=-7>

In [90]:
sum_of_tensor = tf.math.reduce_sum(D, axis = None, keepdims = False, name = None)
sum_of_tensor

<tf.Tensor: shape=(), dtype=int32, numpy=-17>

In [91]:
D

<tf.Tensor: shape=(2,), dtype=int32, numpy=array([ -7, -10], dtype=int32)>

In [94]:
# create a random tensor with values between 0 and 100 of size 50 
E = tf.constant(np.random.randint(0, 100, size = 50))
E

<tf.Tensor: shape=(50,), dtype=int64, numpy=
array([65, 47,  3, 65,  4, 52,  6, 96, 65, 64, 70, 99, 87, 84, 53, 50, 93,
        4, 42, 17, 95, 41, 40, 72, 67, 70, 61, 37, 23, 51, 64, 48, 14, 46,
        6, 75,  5, 11, 63, 54, 10,  0, 44,  9, 38, 11, 55, 53, 81,  2])>

In [95]:
tf.size(E), E.shape, E.ndim

(<tf.Tensor: shape=(), dtype=int32, numpy=50>, TensorShape([50]), 1)

In [98]:
# With keepdims = True 
sum_of_E = tf.math.reduce_sum(E, axis = None, keepdims = True, name = None)
sum_of_E

<tf.Tensor: shape=(1,), dtype=int64, numpy=array([2312])>

In [99]:
# with keepdims = False 
sum_of_E = tf.math.reduce_sum(E, axis = None, keepdims = False, name = None)
sum_of_E

<tf.Tensor: shape=(), dtype=int64, numpy=2312>

In [100]:
tf.reduce_min(E)

<tf.Tensor: shape=(), dtype=int64, numpy=0>

In [101]:
tf.reduce_max(E)

<tf.Tensor: shape=(), dtype=int64, numpy=99>

In [102]:
tf.math.reduce_mean(E)

<tf.Tensor: shape=(), dtype=int64, numpy=46>

In [ ]:
# find variance and standard deviation 

In [105]:
E = tf.cast(E, dtype = tf.float64) 

In [106]:
tf.math.reduce_std(E, axis = None, keepdims = False, name = None)

<tf.Tensor: shape=(), dtype=float64, numpy=28.979689439329746>

In [109]:
import tensorflow_probability as tfp 

In [111]:
tfp.stats.variance(E, sample_axis = 0, keepdims = False, name = None)

<tf.Tensor: shape=(), dtype=float64, numpy=839.8224>

In [112]:
tf.math.reduce_variance(E)

<tf.Tensor: shape=(), dtype=float64, numpy=839.8224>